### PR TITLE
[java] Update many java rules

### LINKED ISSUE #342: [java] AccessorMethodGeneration: Name clash with another public field not properly handled
**Rule Set:** AccessorMethodGeneration (Java - Design), PMD 5.5.5

**Description:**

It appears that the rule AccessorMethodGeneration fails to resolve name clash with public field in another class with identical name.

**Code Sample demonstrating the issue:**

```java
public class ClassA {

    public String field;

}
```

```java
public class ClassB {

    private String field;

    private class ClassC {

        public void doSomething() {
            ClassA a = new ClassA();
            a.field = "abcd"; // should not be a violation here but reported as one
        }

    }

}
```

Changing `ClassB#field` to e.g. `ClassB#field0` removes the violation.

**Running PMD through:** Gradle

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Update many (104) Java rules to pmd 7. There are still 179 of them

* This introduces a few new xpath functions, which are documented.
* Use a common utility class to share logic between rules (JavaRuleUtil), or to store procedures that are not really rule-specific
* Some rule base classes have been removed. In particular, AbstractJunitRule is turned into ~JUnitRuleUtil~ TestFrameworksUtil
* Some rules have been converted to XPath (which is why they're shown as deleted).
* I tried not to change the definition of rules. I implemented #2893 though

## Related issues

- Refs #2701 
- Fixes #342, #807, #1005, #1212, #1224, #1422, #1565, #1969, #1998, #2147, #2528, #2538, #2542, #2577, #2587, #2712, #2716, #2806, #2880, #2883

## Ready?

- [x] Added unit tests for fixed bug/feature
  - [x] Unit tests for new XPath functions are missing (done in #2917) 
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by adangel** [comment]
Checkstyle Diff Report: https://chunk.pmd-code.org/pmd/dec477f7-e299-49ea-b153-d589978ac619/checkstyle/index.html

**Comment by oowekyala** [comment]
Here's another report, which is more easily searchable, and includes the latest bugfixes of this branch:
* [checkstyle](https://chunk.pmd-code.org/pmd/0fa6f0ea-7425-4908-8ce1-355c6adf8205/diff/checkstyle/index.html)
* [spring](https://chunk.pmd-code.org/pmd/0fa6f0ea-7425-4908-8ce1-355c6adf8205/diff/spring-framework/index.html)

**Comment by oowekyala** [comment]
I think all obvious FPs/FNs have been fixed on this branch and this is ready to be merged.

**Comment by oowekyala** [comment]
Unmergeable until #2917 is merged

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/CompareObjectsWithEquals.xml [hunk 1]
```diff
@@ -210,4 +210,63 @@ public class CompareWithEqualsTest {
 }
         ]]></code>
     </test-code>
+    <test-code>
+        <description>comparing in equals method should be ok</description>
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/CompareObjectsWithEquals.xml:214]: here

**Pull request merged**